network_analysis/find-connected-features-in-utility-network/src/main/java/com/esri/samples/find_connected_features_in_utility_network/FindConnectedFeaturesInUtilityNetworkController.java

```diff
@@ -0,0 +1,503 @@
+              } else {
+                utilityElementGroups.get(networkSourceName).add(utilityElement);
+              }
+            });
+
+            // get the feature layer for the utility element
+            utilityElementGroups.forEach((networkSourceName, utilityElements) -> {
+
+              // load each element's feature table manually. Note this is a workaround for some features that are returned by the trace not being selected properly
+              FeatureTable utilityElementsFT = utilityElements.get(0).getNetworkSource().getFeatureTable();
+              utilityElementsFT.loadAsync();
+              utilityElementsFT.addDoneLoadingListener(()->{
+                if (utilityElementsFT.getLoadStatus() == LoadStatus.LOADED) {
+
+                  // get the layer for the utility element
+                  FeatureLayer layer = (FeatureLayer) mapView.getMap().getOperationalLayers().get(0);
+                  if (layer == null) {
+                    return;
+                  }
+
+                  if (layer.getFeatureTable().getTableName().equals(networkSourceName)) {
+
+                    // convert the elements to features to highlight the result
+                    ListenableFuture<List<ArcGISFeature>> fetchUtilityFeaturesFuture = utilityNetwork.fetchFeaturesForElementsAsync(utilityElements);
+                    fetchUtilityFeaturesFuture.addDoneListener(() -> {
+                      try {
+                        List<ArcGISFeature> features = fetchUtilityFeaturesFuture.get();
+                        // select all the features to highlight them
+                        features.forEach(layer::selectFeature);
+
+                        // update the status label
+                        statusLabel.setText("Trace completed.");
+
+                        // enable the UI
+                        enableUI();
+
+                      } catch (InterruptedException | ExecutionException e) {
+                        new Alert(Alert.AlertType.ERROR,
+                          "Error fetching the corresponding features for the utility elements.").show();
+                      }
+                    });
+                  }
+                } else {
+                  new Alert(Alert.AlertType.ERROR, "Error loading Utility Element Feature Layer.").show();
+                }
+              });
+            });
+
+          }
+        } else {
+          statusLabel.setText("Trace failed.");
+          new Alert(Alert.AlertType.ERROR, "Trace result not a utility element.").show();
+        }
+
+        // enable the UI
+        enableUI();
+
+      } catch (InterruptedException | ExecutionException e) {
+        statusLabel.setText("Trace failed.");
+        new Alert(Alert.AlertType.ERROR, "Error running utility network connected trace.").show();
+      }
+    });
+  }
+
+  /**
+   * Enables both buttons and hides the progress indicator.
+   */
+  private void enableUI() {
```

Comment:
this could be why the progress indicator disappears before the trace is complete? If you remove it from here then the method could be renamed "enableButtonInteraction" or similar, "enableUI" isn't too clear from its name what it does.